apm-agent-plugins/apm-cassandra/apm-cassandra3-plugin/src/main/java/co/elastic/apm/agent/cassandra3/Cassandra3Instrumentation.java

```diff
@@ -0,0 +1,153 @@
+        return named("com.datastax.driver.core.SessionManager");
+    }
+
+    @Override
+    public ElementMatcher.Junction<ClassLoader> getClassLoaderMatcher() {
+        // 2.x-3.x
+        return CustomElementMatchers.classLoaderCanLoadClass("com.datastax.driver.core.BoundStatement");
+    }
+
+    /**
+     * {@link com.datastax.driver.core.SessionManager#executeAsync(Statement)}
+     */
+    @Override
+    public ElementMatcher<? super MethodDescription> getMethodMatcher() {
+        return named("executeAsync")
+            .and(returns(named("com.datastax.driver.core.ResultSetFuture")))
+            .and(takesArgument(0, named("com.datastax.driver.core.Statement")));
+    }
+
+    @Override
+    public Collection<String> getInstrumentationGroupNames() {
+        return Collections.singletonList("cassandra");
+    }
+
+    @Override
+    public Class<?> getAdviceClass() {
+        return Cassandra3Advice.class;
+    }
+
+    public static class Cassandra3Advice {
+
+        /**
+         * {@link Host#getSocketAddress()} has been added in 2.0.2
+         */
+        private static final boolean HAS_HOST_SOCKET_ADDRESS;
+        private static final CassandraHelper cassandraHelper = new CassandraHelper(GlobalTracer.get());
+
+        static {
+            boolean hasHostSocketAddress = false;
+            try {
+                Class.forName("com.datastax.driver.core.Host").getMethod("getSocketAddress");
+                hasHostSocketAddress = true;
+            } catch (ReflectiveOperationException | LinkageError ignore) {
+            }
+            HAS_HOST_SOCKET_ADDRESS = hasHostSocketAddress;
+        }
+
+
+        @Nullable
+        @Advice.OnMethodEnter(suppress = Throwable.class, inline = false)
+        public static Object onEnter(@Advice.Argument(0) Statement statement) {
+            return cassandraHelper.startCassandraSpan(getQuery(statement), statement instanceof BoundStatement);
+        }
+
+        @Nullable
+        private static String getQuery(Statement statement) {
+            if (statement instanceof SimpleStatement) {
+                return ((SimpleStatement) statement).getQueryString();
+            } else if (statement instanceof BoundStatement) {
+                return ((BoundStatement) statement).preparedStatement().getQueryString();
+            }
+            return null;
+        }
+
+        @Advice.OnMethodExit(suppress = Throwable.class, onThrowable = Throwable.class, inline = false)
+        public static void onExit(@Advice.Thrown @Nullable Throwable thrown,
+                                  @Advice.Return ResultSetFuture result,
+                                  @Nullable @Advice.Enter Object spanObj) {
+            if (!(spanObj instanceof Span)) {
+                return;
+            }
+            final Span span = (Span) spanObj;
+            span.captureException(thrown).deactivate();
+            Futures.addCallback(result, new FutureCallback<ResultSet>() {
```

Comment:
I've tested with a bunch of versions and this was quite stable. However, there's no stable API to get the direct executor, so we'd have to implement that in the Cassandra module if we wanted to directly implement the Runnable. I tried to keep the complexity of this module low.

Reply:
Yes, this simplicity is very appealing. I agree that implementing it is a worse option. 
My suggested alternative was wrapping the Future with pooled futures. If it is simple enough, we can do that (and save two allocations per query), otherwise this is good enough.